test/integration/test/philips_light_test.js

```diff
@@ -0,0 +1,114 @@
+var header = new Config('./lib/config/header.js').get('header');
+
+describe("Initiate the connection with foxbox as Philips Hue hub",function(){
+  var credential = config.get('credential'); 
+  var FOXBOX_STARTUP_WAIT_TIME_IN_MS = 5000;
+  var lightinfo; // to store the light status received from foxbox
+  var foxbox_process;
+  
+  before(function(done){
+    this.timeout(500000);
+
+    const hue_location = "localhost:" + config.get('philips_hue.port');
+    nupnp_server.start(config.get('nupnp_server.id'),hue_location,config.get('nupnp_server.port'));
+    
+    philipshue_server.setup(config.get('philips_hue.port'));
+
+    philipshue_server.turnOffLight(1);
+    philipshue_server.turnOffLight(2);
+    philipshue_server.turnOffLight(3);
+
+    // kick off the foxbox within 10 seconds
+    //remove_sqlite = spawn('rm',['users_db.sqlite']);
+    fs.unlink('users_db.sqlite');
+    foxbox_process = spawn('../../target/debug/foxbox', 
+      ['-c', 'philips_hue;nupnp_url;http://localhost:'+ config.get('nupnp_server.port')+'/']);
+    setTimeout(done, FOXBOX_STARTUP_WAIT_TIME_IN_MS);  
+  });
+  
+  after(function(){
+    foxbox_process.kill();
+  });
+  
+  it("create a new credential",function(){
+
+    return chakram.post(config.get('foxbox.url') + "/users/setup",credential)
+    .then(function(setupResp){
+      expect(setupResp).to.have.status(201);
+    });
+  });
+
+  it("login to foxbox",function(){
+    var key = credential['username']+":"+credential['password'];
+    var encoded_cred = new Buffer(key).toString('base64');
+
+    header['Authorization'] = 'Basic ' + encoded_cred;  // supply the credential used in previous test
+    
+    return  chakram.post(config.get('foxbox.url') + "/users/login",null,{'headers' : header})
+    .then(function(loginResp){
+      expect(loginResp).to.have.status(201);
+    });
+  });
+
+  describe ("once logged in", function () {
+    before(function() {
+      return chakram.post(config.get('foxbox.url') + "/users/login",null,{'headers' : header})
+      .then(function(resp){
+       header['Authorization'] = 'Bearer ' + resp.body.session_token;
+       chakram.setRequestDefaults({'headers': header});
+     });
+    });
+
+    it("get a list of available philips services, check 3 bulbs are registered",function(){   
+
+      return chakram.get(config.get('foxbox.url') + "/services/list")
+      .then(function(listResponse) {
+        expect(eval(listResponse.body).length == 3);
+        expect(listResponse).to.have.status(200);
+      });
+    });
+
+    describe("once the lights are known", function(){
+      before(function() {
+        return chakram.get(config.get('foxbox.url') + "/services/list")
+        .then(function(listResponse) {
+          lightinfo = listResponse.body;
+        });
+      });
+
+      // Currently, there is no mapping between the foxbox issues id and the philips hue id
+      // until the tag feature is implemented
+      it("Turn on all lights", function(){
```

Comment:
hmm, when I tried this, it looks like the test finishes before waiting for expect statement evaluation, even when there is a return call for chakram API use.  I guess the async nature of expect doesn't like foreach()? 

Another thing is that the order of lightinfo is random, and they do not correspond to the philips hub ID.  so lightinfo[0] may be lightStatus[1] or [2].  That's why I had to separately check all lights are on, in case foxbox generated duplicate IDs for different lights.